kafka/server: add metrics and config for consumer lag reporting

IoannisRP opened the pull request:
Implements: [https://redpandadata.atlassian.net/browse/CORE-8914](https://redpandadata.atlassian.net/browse/CORE-8914)

Introduce "enable_consumer_group_metrics" which controls whether the group, partition and consumer lag metrics are active. This can be changed without needing a restart.

Introduce the metrics scaffolding needed to have metrics that can be enabled/disabled at runtime.

| Metric | Type | Description | Labels | Aggregation labels |
| -- | -- | -- | -- | -- |
| `redpanda_kafka_consumer_group_lag_max` | gauge | Maximum consumer group lag across topic-partitions | `group`, `shard` |  |
| `redpanda_kafka_consumer_group_lag_sum` | gauge | Sum of consumer group lag for all topic-partitions | `group`, `shard` |  |
## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

IoannisRP committed:
kafka/server: move all group probe generation logic into a single function

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--7db2783b-a53f-44ee-b307-f69b9ab7e733-->
<!--TEST RESULTS-->
<!--930ff50b-4153-4b07-a4b8-724308ee0eed-->
<!--TEST RESULTS-->
<!--49dd70f9-e6d3-4f80-8312-69d09bb9c8a4-->
<!--TEST RESULTS-->
<!--c2b631d0-a04f-4fdf-bd5c-397c20085b9c-->
<!--TEST RESULTS-->
<!--909370bd-33c4-46d9-bd4b-28da986826d7-->
<!--TEST RESULTS-->
<!--9a7b3f63-0561-4dbf-87c6-79f912a3df4b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61359">build#61359</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61359#0194b2e4-6b16-498b-a3a6-1735b6488617 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61359#0194b2e9-78d9-4746-a266-103519385a06 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61509">build#61509</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61509#0194cc8d-8071-43c3-93d5-4cffec6c21a1 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/61509#0194cc8d-8071-4d90-8b79-c330b1e751c3 | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62101">build#62101</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62101#01952886-e58c-4e09-a821-7b29cfc21f8e | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62101#01952886-e58e-48dd-ad02-2e4d003944e3 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62101#01952886-e58d-490f-86be-61f7a06cfe43 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62188">build#62188</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.archival_test.ArchivalTest.test_all_partitions_leadership_transfer.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62188#0195387f-fd4e-4cbe-ba0d-ef681e56e324 | FLAKY | 1/2 |
| rptest.tests.datalake.mount_unmount_test.MountUnmountIcebergTest.test_simple_remount.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62188#01953899-8830-405e-b6e6-3ff87c109dd6 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62188#0195387f-fd4e-4d41-a042-7c1426bed80e | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62188#0195387f-fd4e-4d41-a042-7c1426bed80e | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62278">build#62278</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cluster_config_test.ClusterConfigTest.test_valid_settings | ducktape | https://buildkite.com/redpanda/redpanda/builds/62278#01953e40-e0cf-4dd2-8127-f8bffd5a0543 | FAIL | 0/20 |
| rptest.tests.cluster_config_test.ClusterConfigTest.test_valid_settings | ducktape | https://buildkite.com/redpanda/redpanda/builds/62278#01953e5b-49d6-49f0-a610-c18c2e50213a | FAIL | 0/20 |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62278#01953e40-e0cd-47b0-84a7-2196c8d983b9 | FLAKY | 1/3 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62278#01953e40-e0cf-4dd2-8127-f8bffd5a0543 | FLAKY | 1/2 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_relaxed_acks.write_caching=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/62278#01953e40-e0cf-4dd2-8127-f8bffd5a0543 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62404">build#62404</a></summary>


| test_id                                                                                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.mount_unmount_test.MountUnmountIcebergTest.test_simple_unmount.cloud_storage_type=CloudStorageType.S3                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62404#01954779-7ba7-4d9b-aedd-096c242b6706 | FLAKY         | 1/2      |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=node_add.test_mode=TestMode.TIRED_STORAGE.cleanup_policy=compact.delete | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62404#01954779-7ba9-419f-b331-aed451fd58df | FLAKY         | 1/2      |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.S3                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62404#01954779-7ba6-47f1-a9f1-a4fc36bf9037 | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62404#01954779-7ba9-419f-b331-aed451fd58df | FLAKY         | 1/2      |
</details>
<!--/body-->

michael-redpanda commented on src/v/kafka/server/group.h:
```diff
@@ -965,6 +967,7 @@ class group final : public ss::enable_lw_shared_from_this<group> {
     chunked_hash_map<model::topic_partition, offset_metadata>
       _pending_offset_commits;
     enable_group_metrics _enable_group_metrics;
+    config::binding<bool> _enable_consumer_lag_metrics;
```
nit: you aren't using or assigning this in this commit so it should be moved to a different commit

BenPope replied:
This could be moved to the next commit.

michael-redpanda commented on tests/rptest/tests/consumer_group_test.py:
```diff
@@ -658,6 +658,64 @@ def test_consumer_static_member_update(self):
+                                          static_members=False)
+
+        self.start_producer()
+        # wait for some messages
+        wait_until(
+            lambda: ConsumerGroupTest.group_consumed_at_least(
+                consumers, 50 * len(consumers)), 30, 2)
```
nit: add an error message in `wait_until`

michael-redpanda reviewed (commented):
nice!

BenPope commented on src/v/kafka/server/group_probe.h:
```diff
@@ -137,11 +146,51 @@ class group_probe {
+    // To deregister the metrics the public_metric_groups object needs to be
+    // destroyed. For this reason, a nullable type is being used here.
+    std::optional<metrics::public_metric_groups> _public_metrics_consumer_lag;
+    bool _disable_public_metrics;
```
Why bind this at construction rather than reading the current value?

IoannisRP replied:
Mainly because I needed to have a watcher on it. Only the bindings have an interface for watchers.

BenPope replied:
> Mainly because I needed to have a watcher on it. Only the bindings have an interface for watchers.

I'm not sure I understand, `_disable_public_metrics` isn't watched, right?

IoannisRP replied:
`_disable_public_metrics` needs to be restarted for the new value to take effect. For consumer lag we want to be able to enable/disable without restart.

BenPope replied:
> `_disable_public_metrics` needs to be restarted for the new value to take effect. For consumer lag we want to be able to enable/disable without restart.

Right, but it wouldn't if it you read from it directly, or did a bind to it.

BenPope commented on src/v/kafka/server/group_probe.h:
```diff
@@ -137,11 +146,51 @@ class group_probe {
+             "lag_sum",
+             [this] { return _lag_metrics.sum; },
+             sm::description(
+               "Sum of consumer group lag for all partitions in a group"),
+             labels),
+           sm::make_gauge(
+             "lag_max",
+             [this] { return _lag_metrics.max; },
+             sm::description(
+               "Maximum consumer group lag across all partitions in a group"),
```
Maybe one for docs, but topic-partitions might be more easily understood.

IoannisRP replied:
Like this?
```suggestion
               "Sum of consumer group lag for all topic-partitions"),
             labels),
           sm::make_gauge(
             "lag_max",
             [this] { return _lag_metrics.max; },
             sm::description(
               "Maximum consumer group lag across topic-partitions"),
```

asimms41 replied:
I agree.

BenPope commented on src/v/kafka/server/group_probe.h:
```diff
@@ -137,11 +146,51 @@ class group_probe {
+             labels)});
+    }
+    void deregister_consumer_lag_metrics() {
+        _public_metrics_consumer_lag = std::nullopt;
```
nitpick:
```suggestion
        _public_metrics_consumer_lag.reset();

BenPope commented on src/v/config/configuration.h:
```diff
@@ -196,6 +196,7 @@ struct configuration final : public config_store {
     property<bool> disable_metrics;
     property<bool> disable_public_metrics;
     property<bool> aggregate_metrics;
+    property<bool> enable_consumer_group_lag_metrics;
```
I wonder if we should have more than a bool here.

I see a use for:
* partition-level metrics (what we have with `redpanda_kafka_consumer_group_committed_offset`)
* consumer_lag (this)
* both

`enable_group_metrics` is hard-coded as on (off in some tests), so it could be wrapped up.

Perhaps `enable_group_metrics` with options `partition`, `group`. Not sure how to spell the option both, if, perhaps, we wanted to add `topic` one day. Maybe the accepted values could be something like `partition|lag`.

michael-redpanda replied:
This sounds like something maybe we can tackle in future metrics rework?

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/76291afc896d552124a9ac86a21f72c1c50d0037..53aa75613fc480e008a254413f28dd9abdbc90a2):
- rebase to dev

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/53aa75613fc480e008a254413f28dd9abdbc90a2..b880e01b27441ace562a1973d687ead85d981d12):
- merged new config to commit that uses it
- removed value read on construction for `_disable_public_metrics`
- added comment in ducktape test `wait_until`

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -582,6 +582,13 @@ configuration::configuration()
+  , enable_consumer_group_lag_metrics(
+      *this,
+      "enable_consumer_group_lag_metrics",
+      "Enable registering metrics for consumer group lag exposed on "
```
```suggestion
      "Enable metrics for consumer group lag exposed on "
```

IoannisRP committed:
kafka/server: Replace hardcoded boolean value with config.

Replace enable_group_metrics hardcoded value with
enable_consumer_group_metrics config, which controls group
and partition metrics independently.

Rework group_probe and group_offset_probe to also be able to dynamically
register and deregister their metrics.

IoannisRP committed:
kafka/server: add consumer lag metrics

Note that this commit contains only the metric infrastructure, i.e. the
probe and the mechanism to dynamically enable/disable these metrics.
A subsequent commit will implement the logic to populate the consumer
lag metrics data.

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/b880e01b27441ace562a1973d687ead85d981d12..1391118d1f8a563ba659de4fe5c81b9e64516ebf):
- Rebase to dev.
- Reworked config and existing probes to be able to control all group, partition and consumer group metrics, independently.

BenPope commented on src/v/config/types.h:
```diff
@@ -151,4 +155,81 @@ inline std::istream& operator>>(std::istream& is, datalake_catalog_type& ct) {
+namespace kafka::group_metrics {
+enum class bitflags : uint8_t { none = 0, group = 1U, partition = 1U << 1U };
+
+inline bitflags operator|(const bitflags lhs, const bitflags rhs) {
+    return bitflags(
+      std::underlying_type_t<bitflags>(lhs)
+      | std::underlying_type_t<bitflags>(rhs));
+}
```
nitpick: we have c++23!
```suggestion
constexpr bitflags operator|(const bitflags lhs, const bitflags rhs) {
    return bitflags(std::to_underlying(lhs) | std::to_underlying(rhs));
}
```

BenPope commented on src/v/config/types.h:
```diff
@@ -151,4 +155,81 @@ inline std::istream& operator>>(std::istream& is, datalake_catalog_type& ct) {
+      | std::underlying_type_t<bitflags>(rhs));
+}
+
+inline void operator|=(bitflags& lhs, const bitflags rhs) { lhs = (lhs | rhs); }
```
nitpick: No reason not to make this `constexpr`, which implies inline:
```suggestion
constexpr void operator|=(bitflags& lhs, const bitflags rhs) {
    lhs = (lhs | rhs);
}
```

BenPope commented on src/v/config/types.h:
```diff
@@ -151,4 +155,81 @@ inline std::istream& operator>>(std::istream& is, datalake_catalog_type& ct) {
+    if (is_partition_enabled(flags)) {
+        ret.emplace_back("partition");
+    }
+    return fmt::format("{}", fmt::join(ret, ","));
```
allocations could be reduced:
```suggestion
    std::array<std::string_view, 2> ret;
    auto end = ret.begin();
    if (is_group_enabled(flags)) {
        *end++ = "group";
    }
    if (is_partition_enabled(flags)) {
        *end++ = "partition";
    }
    return fmt::format("{}", fmt::join(ret.begin(), end, ","));
```

IoannisRP replied:
I replaced `std::string` -> `std::string_view` and added a reservation for 3 objects.

Given the use of this type, I don't think it is worth to sacrifice readability (the array approach) or compilation speed (to bring in some other type (like `boost::static_vector`) to shave off one allocation.

BenPope commented on src/v/config/types.h:
```diff
@@ -151,4 +155,81 @@ inline std::istream& operator>>(std::istream& is, datalake_catalog_type& ct) {
+
+inline void operator|=(bitflags& lhs, const bitflags rhs) { lhs = (lhs | rhs); }
+
+inline bitflags operator&(const bitflags lhs, const bitflags rhs) {
+    return bitflags(
+      std::underlying_type_t<bitflags>(lhs)
+      & std::underlying_type_t<bitflags>(rhs));
+}
+
+inline void operator&=(bitflags& lhs, const bitflags rhs) { lhs = (lhs & rhs); }
+
+inline bool is_group_enabled(const bitflags flags) {
+    return (flags & bitflags::group) != bitflags::none;
+}
+
+inline bool is_partition_enabled(const bitflags flags) {
+    return (flags & bitflags::partition) != bitflags::none;
+}
```
```suggestion
constexpr bitflags operator&(const bitflags lhs, const bitflags rhs) {
    return bitflags(std::to_underlying(lhs) & std::to_underlying(rhs));
}

constexpr void operator&=(bitflags& lhs, const bitflags rhs) {
    lhs = (lhs & rhs);
}

constexpr bool is_group_enabled(const bitflags flags) {
    return (flags & bitflags::group) != bitflags::none;
}

constexpr bool is_partition_enabled(const bitflags flags) {
    return (flags & bitflags::partition) != bitflags::none;
}
```

BenPope commented on src/v/config/configuration.cc:
```diff
@@ -596,6 +597,14 @@ configuration::configuration()
+  , enable_consumer_group_metrics(
+      *this,
+      "enable_consumer_group_metrics",
+      "List of enabled consumer group metrics. Accepted Values: "
```
Might be worth mentioning that the list is comma separated.

BenPope commented on src/v/config/types.h:
```diff
@@ -151,4 +155,81 @@ inline std::istream& operator>>(std::istream& is, datalake_catalog_type& ct) {
+}
+
+inline bitflags from_string(std::string_view sv) {
+    if (sv == "none") {
+        return bitflags::none;
+    }
+
+    bitflags ret{};
+    auto parts = std::views::split(sv, ',');
+    for (const auto& part : parts) {
+        const std::string_view option{part};
```
Is it worth stripping whitespace?

IoannisRP replied:
I don't really know. I was mainly copying the behavior of  `cleanup_policy_bitflags` which doesn't allow for commas.

To me, writing "group ,partition" feels awkward. But there might be a legitimate use case for it.

BenPope replied:
Probably not worth it

BenPope commented on src/v/kafka/server/group.h:
```diff
@@ -633,7 +645,10 @@ class group final : public ss::enable_lw_shared_from_this<group> {
-                std::move(md), _id, tp, _enable_group_metrics));
+                std::move(md),
+                _id,
+                tp,
```
Use after move?

IoannisRP replied:
Are referring to `tp`? I didn't notice that, it is pre-existing.

It is worth a separate PR, so it can be backported.

BenPope replied:
> Are referring to tp? I didn't notice that, it is pre-existing.

Yes.

> It is worth a separate PR, so it can be backported.

Yes, it's probably worth backporting. I'll leave the mechanism up to you.

BenPope commented on src/v/kafka/server/group.h:
```diff
@@ -648,7 +663,10 @@ class group final : public ss::enable_lw_shared_from_this<group> {
-                std::move(md), _id, tp, _enable_group_metrics));
+                std::move(md),
+                _id,
+                tp,
```
Use after move?

IoannisRP replied:
same

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/1391118d1f8a563ba659de4fe5c81b9e64516ebf..fc54813af7d5d33c06478c482ba9c030daebe1e5):

- rebased to dev to solve merge conflict

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/fc54813af7d5d33c06478c482ba9c030daebe1e5..f0e44baa897e883c51745b41dd3e488dbb45cd69):
- Address review comments

michael-redpanda commented on src/v/config/configuration.cc:
```diff
@@ -596,6 +597,14 @@ configuration::configuration()
+      "Values: `group`, `partition`, `none`",
+      {.needs_restart = needs_restart::no, .example = "group,partition"},
+      config::kafka::group_metrics::bitflags::group
+        | config::kafka::group_metrics::bitflags::partition)
```
suggestion: Make this optional

It's confusing that you can technically have `group,partition,none` as a valid entry.  Instead this should maybe be an optional. with just `group` and `partition` as options.

IoannisRP replied:
i am not sure what optional refers to in this context. Do we have/Could you point to another config that behaves the way you would like to see this one behaving it?

note that if the user where to set the value to `group,none` (or any other combination with `none`), and then query the value, they get back just `group`. I don't know if it would help, but it is trivial to add an extra line of code to explicitly render `none` invalid in any combination other than stand alone. I just didn't think it makes any practical difference.

michael-redpanda replied:
RIght so there's a reconciliation error here - if something set's the config to `group,none` and then gets back `none` it'll think it's config didn't take and rerun.  I know it's a contrived exampled but it demonstrates the usability issue here.

Using an optional negates the need to have `none` and to further special case the `none` option.

IoannisRP replied:
I think i am missing some context here. Let's have an offline chat about it.

michael-redpanda reviewed (changes requested):
I think the config property needs to be slightly reworked to improve user experience

IoannisRP commented on src/v/config/types.h:
```diff
@@ -193,4 +198,98 @@ operator>>(std::istream& is, datalake_catalog_auth_mode& cam) {
+    auto parts = std::views::split(sv, ',');
+    for (const auto& part : parts) {
+        const std::string_view option{part};
+        const std::optional<bitflags> f
+          = string_switch<std::optional<bitflags>>(option)
+              .match("none", bitflags::none)
+              .match("group", bitflags::group)
+              .match("partition", bitflags::partition)
+              .match("consumer_lag", bitflags::consumer_lag)
+              .default_match(std::nullopt);
+        if (!f.has_value()) {
+            throw std::runtime_error(
+              fmt::format("Invalid value \"{}\"", option));
+        }
```
```suggestion
        const std::optional<bitflags> f
          = string_switch<std::optional<bitflags>>(option)
              .match("none", bitflags::none)
              .match("group", bitflags::group)
              .match("partition", bitflags::partition)
              .match("consumer_lag", bitflags::consumer_lag)
              .default_match(std::nullopt);
        if (!f.has_value()) {
            throw std::runtime_error(
              fmt::format("Invalid value \"{}\"", option));
        }
        if (f.value() == bitflags::none) {
            throw std::runtime_error("explaining why none can only be used in isolation here...");
        }
```
@michael-redpanda this is the change i refer to in the above comment

michael-redpanda replied:
Switching to optional negates the need for `none`

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/f0e44baa897e883c51745b41dd3e488dbb45cd69..afb618b2f2774c04cb7ba5b98e69c874499d302d):
- rework config to be a list of strings instead of the flag-based approach. The reason for this change was that the previous approach didn't remember the input string (i.e. `partition,group,none` would be reported back as `group,parition`), which might cause issues down the line with tooling that expects the setting to directly match the input.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62278-->
<!--6b945a6b-556f-4d69-9834-5aaddc780010-->
<!--9ba555ad-c5a1-44c5-8fb9-8bfdd9f5f0a3-->
<!--title-->
## Retry command for [Build#62278](https://buildkite.com/redpanda/redpanda/builds/62278)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/cluster_config_test.py::ClusterConfigTest.test_valid_settings
```
<!--/body-->

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/afb618b2f2774c04cb7ba5b98e69c874499d302d..9da0d13cfc100a49b7eacc0f34a721d4c78c9e2b):
- Update a failing dt test due to new config

BenPope commented on src/v/config/types.h:
```diff
@@ -12,7 +12,12 @@
 #include <ostream>
+#include <ranges>
+#include <stdexcept>
+#include <utility>
```
Are these new includes still needed?

BenPope commented on src/v/kafka/server/group_probe.h:
```diff
@@ -24,6 +25,40 @@
+        if (contains("partition")) {
+            ret.partition = true;
+        }
+        return ret;
```
nitpick:
```suggestion
        return enabled_metrics{
          .group = std::ranges::contains(metrics, "group"),
          .partition = std::ranges::contains(metrics, "partition")};
```

BenPope commented on src/v/kafka/server/group_probe.h:
```diff
@@ -24,6 +25,40 @@
+    };
+    return compare(&enabled_metrics::group)
+           && compare(&enabled_metrics::partition);
+}
```
nitpick:
```suggestion

    friend constexpr auto
    operator<=>(const enabled_metrics&, const enabled_metrics&)
      = default;
};
```

BenPope replied:
Making it a hidden friend was intentional, but the change is fine.

BenPope reviewed (commented):
Looks great, just cleanups and nits

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/9da0d13cfc100a49b7eacc0f34a721d4c78c9e2b..0897e7eb4f4e3a2216f9801fd0e08c3bc2f00f10):
- Addressed comments

michael-redpanda reviewed (approved):
🔥

michael-redpanda commented on src/v/kafka/server/group.h:
```diff
@@ -227,18 +228,28 @@ class group final : public ss::enable_lw_shared_from_this<group> {
+          , probe(metadata.offset)
+          , enable_group_metrics(std::move(_enable_group_metrics)) {
+            const auto metrics_registration = [this, group_id, tp]() {
+                if (enable_group_metrics().partition) {
+                    probe.register_metrics(group_id, tp);
+                    probe.register_public_metrics(group_id, tp);
+                } else {
+                    probe.deregister_metrics();
+                    probe.deregister_public_metrics();
+                }
```
I'm wondering if this is the problem with the double registration.

We start with `group,partition` so the partition metrics get registered.  Then it changes to `partition` only, this watch callback will get called and because `enable_group_metrics().partition` returns `true`, we see a double registration here.

@IoannisRP @BenPope

michael-redpanda replied:
maybe not... i see there's a `std::optional` check
